src/blob/index/core.rs

```diff
@@ -295,28 +298,51 @@ where
         }
     }
 
-    async fn get_all(&self, key: &K) -> Result<Option<Vec<RecordHeader>>> {
-        match &self.inner {
+    async fn get_all(&self, key: &K) -> Result<ReadResult<Vec<RecordHeader>>> {
```

Comment:
My thoughts on the `get_all` function logic. At the `Storage` level, we need not only undeleted data, but also a `deleted` marker, which will allow to stop scanning old BLOBs. Unfortunately, that cannot be expressed with `ReadResult<>` type.
I think the `get_all` function can be split into 2 functions (on all levels: index, blob, storage):
1. `async fn get_all(&self, key: &K) -> Result<Vec<RecordHeader>>`: this version preserves the old behavior and returns only non-deleted records. If there are no non-deleted records, then it simply returns empty vector
2. `async fn get_all_with_deleted(&self, key: &K, go_beyond_deleted: bool) -> Result<Vec<RecordHeader>>`: this version returns all non-deleted records and then the record with `deleted` flag (if it was deleted). When `go_beyond_deleted == true` it also returns all records after the deleted one

Example for key `1`: v2, del, v1
1. get_all(1): `v2`
2. get_all_with_deleted(1, false): `v2, del`
3. get_all_with_deleted(1, true): `v2, del, v1`

In this way we can support all scenarios + this allow to stop scan on the deletion marker on `Storage` level

Reply:
I suppose the order should be reversed? `v1, del, v2`

Reply:
Not sure why we need those. I've changed `read_all` on storage so it scans blobs and returns early if necessary. Is it necessary to search for records after deletion marker though?

Reply:
For the user: we need a function that gets clear results without any special entries indicating deletion.
For internal needs (early stop on a removed record): we need a function that returns at least the first deletion marker.
The third option, which returns all records, ignoring deletion markers, is needed just in case (its implementation is trivial)

Reply:
But isn't deletion supposed to be non-recoverable from the user's perspective?

Reply:
Hm. Yes, that's a good argument.
Ok. Then I think the third option is not needed. We need just a 2 functions: 
1. `async fn get_all(&self, key: &K) -> Result<Vec<RecordHeader>>`
2. `async fn get_all_with_deletion_marker(&self, key: &K) -> Result<Vec<RecordHeader>>`

`get_all` can call `get_all_with_deletion_marker` where possible and remove the last item if it is a deletion marker

Reply:
See the comment below for an example describing why we need the second function and why `ReadResult<>` does not work for the `get_all` case